### PR TITLE
C++ Interop: Reserved identifiers

### PR DESCRIPTION
**Pull request by JamesJCode**
This proposal reserves a space of identifiers to enable smooth and error-free C++ interop, and to provide a reserved namespace for internal Carbon implementation specifics.

### CODE REVIEW

File: docs/design/README.md [hunk 1]
```diff
@@ -1827,6 +1827,13 @@ and starts with a letter. We intend to follow
 identifier characters, but a concrete set of valid characters has not been
 selected yet.
 
+The following names are reserved for internal Carbon implementation specifics,
+and to allow seamless interoperability with C++ codebases:
+
+-   Names with a double underscore anywhere are reserved;
+-   Names that begin with an underscore followed by an uppercase letter are
+    reserved.
+
```
Comment (josh11b) [inline docs/design/README.md:1830]: This feels a bit detailed for the overview. Can we make this a summary with fewer details, and link to the detailed design?

File: proposals/p2029.md [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+<!-- tocstop -->
+
+## Problem
+
+C++ reserves certain sets of identifiers for implementation-specific use. Carbon
+must ensure it respects these reserved identifiers to enable seamless and
+error-free interoperability with C++. Equally, it would be prudent to define a
+set of identifiers which are reserved for internal Carbon implementation usage.
+
+## Background
```
Comment (josh11b) [inline proposals/p2029.md:31]: What about C++ keywords?

File: docs/design/lexical_conventions/words.md [hunk 3]
```diff
@@ -82,6 +83,15 @@ The following words are interpreted as keywords:
 -   `where`
 -   `while`
 
+## Identifiers
```
Comment (josh11b) [inline docs/design/lexical_conventions/words.md:86]: As a general rule, I don't think we want a separate detailed design that just repeats what is in the overview. In this case, I think it would be reasonable to move overview content here, but another option is to make the overview itself the authority without any separate detailed design document.

File: docs/design/README.md (hunk 1 above)
Comment (JamesJCode) [reply docs/design/README.md:1830]: Fixed - link to PR was already in the References section, so just wording adjusted.

File: proposals/p2029.md (hunk 2 above)
Comment (JamesJCode) [reply proposals/p2029.md:31]: In the Background section, I link  to the (stalled) PR which discusses this: https://github.com/carbon-language/carbon-lang/pull/93

File: docs/design/lexical_conventions/words.md (hunk 3 above)
Comment (JamesJCode) [reply docs/design/lexical_conventions/words.md:86]: Happy either way, but I did see previous discussions about a concern that design docs were getting out of sync with proposal PRs. My concern was whether this was a lexical issue or a semantic one (i.e. we could emit a separate token, say `RESERVED_IDENTIFIER`, or whether semantic analysis later should check all identifiers). I put it here for now to at least trigger both these discussions.

File: proposals/p2029.md [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+Carbon toolchains are expected to emit an error when a reserved identifier is
+encountered. This will remove the risk of undefined / unexpected runtime
+behaviour, or name resolution clashes when linking with C++ libraries.
+
+## Details
+
+This proposal makes use of three assumptions:
+
+1. A Carbon package exporting a C++ header will be scoped in a C++ namespace
+   following the package name;
```
Comment (jonmeow) [inline proposals/p2029.md:64]: To explain the behavior I expect:

By default, some (but not all) Carbon names will be exported to a matching C++ namespace. However, this won't include everything (e.g., some Carbon features will preclude availability in C++). Also, there'll probably be some capacity for renaming the exported symbol. Off the top of my head, I have two reasons, both based on migration of C++ code: (1) there may be an API in C++'s global namespace that's being replaced; (2) Carbon's naming and package idioms may not match how C++ code was written.

So what you're describing here is a default behavior, but not the only behavior.

File: proposals/p2029.md [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+2. All C++ imports to Carbon will exist within the package `Cpp` which
+   represents the root of the C++ global namespace.
+3. The default Carbon package can not be exported to C++, following the
+   [no other library can import this library even from within the default package](/docs/design/README.md#files-libraries-packages)
+   rule.
```
Comment (jonmeow) [inline proposals/p2029.md:69]: I think this is a misunderstanding. As written, I think it's confusing packages and libraries, and maybe you meant to say:

"_A package's default library_ can not be exported to C++, following the [no other library can import _the default library_ even from within the _same package_](/docs/design/README.md#files-libraries-packages) rule."

As presently discussed, no other library _within the current package_ can import the default library. That is, a library written `package Foo api;` cannot be imported from elsewhere within the package `Foo` because `import Foo;` would be the correct syntax, but `import library <name>` is _required_ when a library imports from itself.

In other words, other packages can import the default library of a package. And similarly, the default library can be exported to C++.

File: proposals/p2029.md [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+reservation.
+
+## Rationale
+
+This proposal supports the
+[Interoperability with and migration from existing C++ code](/docs/project/goals.md#interoperability-with-and-migration-from-existing-c-code)
+goal. It also provides a reserved identifier space for Carbon internal
+implementation specifics.
+
+## Alternatives considered
```
Comment (jonmeow) [inline proposals/p2029.md:86]: My leaning would be to not do anything here. If desired, much as some Carbon features cannot be directly exposed to C++ ([requiring bridge code](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/interoperability/philosophy_and_goals.md#minimize-bridge-code)), code with C++-reserved identifiers could be treated similarly.

An advantage of that approach is that it avoids constraining Carbon's expressiveness based on what can be expressed in C++. While I think that may not be noticed short-term, long-term I think it could hinder Carbon's evolution.

@josh11b mentions C++ keywords above: I see those as a more significant evolution issue. It's worth noting that in the future C++ may add more keywords that Carbon doesn't. If the conclusion is that Carbon should reserve any reserved name in C++, it puts Carbon on the hook for preventing identifiers that match keywords not just in _current_ C++ standards, but _future_ C++ standards: a change to _C++_ could render Carbon code invalid, which I would lean away from.

Unlike C++, Carbon can also place identifiers that it may need in a package that it owns (the core library, or similar). It can also create anonymous identifiers. It shouldn't need to reserve identifiers in every package for this purpose.

A disadvantage of this approach is that, yes, exporting Carbon names that use C++ language-reserved identifiers will require a manual rename. However, my own leaning is that this is a reasonable disadvantage and that developers trying to write C++-interoperable code won't be particularly surprised by the consequence. This avoids the concerns of an automatic rewrite causing a conflict.

File: proposals/p2029.md [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+It was considered to not place any user-visible restrictions on identifiers, and
+handle conflict resolution or avoidance at interop boundaries. This was
+dismissed as this would require automatic re-writing of reserved-in-C++
+identifiers, deferring identification of name resolution issues to link time, or
```
Comment (jonmeow) [inline proposals/p2029.md:91]: Wouldn't most name resolution issues be at compile time (nothing to find), rather than link time?

**Review by jonmeow — COMMENTED** [review]
Note, I'd prefer an alternative to be adopted: while we probably want to declare some reserved names, I think adopting C++'s underscore-based reservation scheme is something Carbon can avoid due to namespacing.

File: proposals/p2029.md (hunk 4 above)
Comment (JamesJCode) [reply proposals/p2029.md:64]: Is there any documentation on what the non-default behaviour could look like? Or does this proposal need to wait for futher interop design / evolution work?

File: proposals/p2029.md (hunk 5 above)
Comment (JamesJCode) [reply proposals/p2029.md:69]: I was referring to these two elements of the linked doc, which are explicitly about default package export or import (although the second point does link the two defaults):

- If the package name is omitted, as in `package library "Main" api;`, the file contributes to the default package. No other package may import from the default package.
- If a file has no package declaration at all, it is the `api` file belonging to the default package and default library. This is particularly for tests and smaller examples. No other library can import this library even from within the default package. It can be split across multiple `impl` files using a `package impl;` package declaration.

I was thinking about whether an un-named package could be exported to C++, and what namespace implications that could have (i.e. if it were possible, would it be to the root of the global namespace, and therefore do we need to worry about collisions?) Grateful to know if I have misunderstood something here? The linked title doesn't make this clear though.

File: proposals/p2029.md (hunk 6 above)
Comment (JamesJCode) [reply proposals/p2029.md:86]: OK, does this need to be a question for leads to work out the approach? On Discord there was interest in a short proposal to cover this approach, but happy to bin it if the collective view is otherwise.

On the reserved C++ keywords issues, the stalled PR #93 I link in the background section suggestions options such as warning programmers if they have used a reserved C++ keyword. Maybe this PR needs revisiting to nail this behaviour down?

File: proposals/p2029.md (hunk 7 above)
Comment (JamesJCode) [reply proposals/p2029.md:91]: What I meant by this was the (probably small) risk of finding when linking that there were two symbols with the same name (and / or signature) at link time due to C++ implementation specifics not exposed in the imported-in-Carbon C++ headers clashing with Carbon user-specified identifiers.

File: proposals/p2029.md (hunk 2 above)
Comment (josh11b) [reply proposals/p2029.md:31]: I think Carbon will have some syntax like Rust's `r#` "raw identifier" syntax for saying "treat this Carbon keyword as an identifier", as discussed in #93 . This will allow C++ APIs using Carbon keywords to be usable from Carbon. The bigger problem is using Carbon APIs that use C++ keywords from C++.

This problem might be out of scope for this proposal, but it seems very related (not using things in Carbon that aren't legal in C++), and worth mentioning. I wonder if there might be a single solution that works for things with underscores as well as C++ keywords. Possibly an "export to C++ as X" annotation?

File: proposals/p2029.md (hunk 5 above)
Comment (jonmeow) [reply proposals/p2029.md:69]: Thanks! I'd missed that this was added to the readme. It's provisional and not really part of the design, so I haven't thought about it much -- you're probably right.

File: proposals/p2029.md (hunk 7 above)
Comment (jonmeow) [reply proposals/p2029.md:91]: Under this argument though, aren't _all_ issues of name resolution deferred until link time because there could be some C++ ODR conflict?

File: proposals/p2029.md (hunk 6 above)
Comment (jonmeow) [reply proposals/p2029.md:86]: The only context I have is from [this discussion](https://discord.com/channels/655572317891461132/999035933959671808/1007120490286035037) -- if there's something more than that, I missed it. chandlerc did mention thinking a reserved space may be helpful, but I'm not sure what the intended use was -- if it was to match C++ or something else (like the lambda use-case being discussed, but I still don't think Carbon should need that).

I don't know what the leads would prefer here. One option may be to just keep going with the proposal and see the response. i.e., it would be reasonable to mention my suggested alternative, offer up advantages and disadvantages, and see what the leads decide. A leads question may yield a similar result with possibly less effort, though they might still want to see the proposal.

Regarding #93, I think that's only mentioning possible uses, it's not trying to decide on them. Even if #93 had been accepted, I don't think it'd give you an answer. It's really a distinct question of whether Carbon leads want Carbon to reserve all C++-reserved identifiers. You could as easily add C++ keywords to the list of reserved identifiers here, which seems consistent with the current rationale of reserving things which would have name resolution clashes with C++.

But FYI both zygoloid and chandlerc probably won't respond this week. I don't know whether @KateGregory would like to comment in their absence. :)

File: proposals/p2029.md (hunk 4 above)
Comment (jonmeow) [reply proposals/p2029.md:64]: [This doc](https://docs.google.com/document/d/1rPJ5GQnd-TLyFULUNNwfl5Mf2fIpX8mR7qcI4SjpQXw/edit?resourcekey=0-A7I-NiJbomij1b0gUW26PQ) has a lot of thought on this, but note that's pretty old and the related PR was deferred.

File: proposals/p2029.md (hunk 2 above)
Comment (JamesJCode) [reply proposals/p2029.md:31]: I can happily throw in a section on this. I think there are a number of options for the keyword issue:

1. Do nothing - users will need to refactor their code if they hit problems when using their libraries in C++
2. Do nothing in regular compilation, error / warn if emitting a C++ header
3. Always error or warn (which is an option presented in #93)
4. Allow an "export to C++ as X" annotation

I note that option 4 can be combined with 1-3 to provide a route to fix which doesn't change the whole Carbon codebase too.

I've seen some discussion flying around on what annotations could look like, but not sure what the current 'state of the art' is; is there a working syntax for now?

File: proposals/p2029.md (hunk 6 above)
Comment (JamesJCode) [reply proposals/p2029.md:86]: So I think it's a non-issue for lambdas (following an approach along the lines you suggested in the discussion), which was where I first raised the question, but it raised the wider discussion about having a reserved identifier space for C++ interop (which could also be used by Carbon if needed in the future - but I did wonder about Carbon using some specific package as you raise too).

But I think your comments on this are intersecting with @josh11b's above, and show that keywords should be included in this proposal - I'll add a section with some options.

File: proposals/p2029.md (hunk 7 above)
Comment (JamesJCode) [reply proposals/p2029.md:91]: I don't think it's *all*, I think it's a fairly small low-risk *some*. As one piece of empirical evidence, I can count on one hand the number of times I've had to deal with this problem when linking against external libraries. I don't think it's a problem we can resolve, but I don't think it's a huge deal.

**Comment by josh11b** [comment]
@carbon-language/carbon-leads A little bit of guidance from leads here would make the path forward clear which would make forward progress easier. In particular this PR is about the question: how much should we make reserved in Carbon since they are reserved in C++ and so using it in Carbon will make it harder to access from C++?

The two main categories of concern are:

- C++ keywords, which change, but not quickly (discussed in [this thread](https://github.com/carbon-language/carbon-lang/pull/2029#discussion_r946121240)).
- Identifiers with specific combinations of underscores and possibly capital letters (discussed in [this thread](https://github.com/carbon-language/carbon-lang/pull/2029/files/60c5f189de1fcf31c97bde3ad2aed7271c70147a#r945980524))

Presumably we will either say these are all reserved in Carbon or there is some way to say "export this Carbon thing as something else in C++." Which one would you prefer? If we have a way to export things to C++ under a different name, do we want to reserve any of these in Carbon?

**Comment by chandlerc** [comment]
> The two main categories of concern are:
> 
> * C++ keywords, which change, but not quickly (discussed in [this thread](https://github.com/carbon-language/carbon-lang/pull/2029#discussion_r946121240)).
> * Identifiers with specific combinations of underscores and possibly capital letters (discussed in [this thread](https://github.com/carbon-language/carbon-lang/pull/2029/files/60c5f189de1fcf31c97bde3ad2aed7271c70147a#r945980524))
> 
> Presumably we will either say these are all reserved in Carbon or there is some way to say "export this Carbon thing as something else in C++." Which one would you prefer?

I think we will really want a very nice, ergonomic way to export a C++-only alias. There are lots of reasons I can imagine here beyond the collisions you mention (dealing with changes to naming conventions, namespace/package name structures, etc). But they'll work just as well to address the collisions you mention.

I also think there is a third option that I would like to include: add a compiler extension to achieve a "raw identifier" in C++. I think we should try to get that into Clang at least, and use it when we can. There are also other users that would benefit from this. It already exists in MSVC's C++/CLI mode ([`__identifier(...)`](https://docs.microsoft.com/en-us/cpp/extensions/identifier-cpp-cli)) and has been requested several times before. We can at least contemplate ensuring this is available in our primary interop host toolchain. The use case here is preserving API aliases for existing code in old language standards.

Personally, I think we should try to have *both* of these things to give ourselves lots of flexibility.

> If we have a way to export things to C++ under a different name, do we want to reserve any of these in Carbon?

After thinking about the above, my guess (but this is just a very weak guess on my part) is that we shouldn't try to use reserved identifier space to avoid these collisions.

But I think there is some utility in having a strong convention about how to write *grammatically* valid identifiers that explicitly don't collide with user code. And similarly, I think its useful to give implementations a space they *can* use if actually rewriting source code is desirable for some reason (even though I don't think this is how we should realistically implement things in practice).

So I would still suggest reserving some space, but more for our own purposes.

I would also suggest having the Carbon reserved names be visually similar and within the reserved space for C/C++. My personal suggestion would just be the extremely simple rule of names starting with `__`. This is much simpler than C++'s rules, but conventional and familiar. It gives us at least some space if useful.

My two cents.

**Comment by JamesJCode** [comment]
FWIW I've basically come to the same conclusion (less the adding raw identifiers to the C++ compiler idea). For exported names, this could look like having something following the attribute syntax such as:

```
@Cpp.ExportAs("CppFunction")
fn CarbonFunction(){...}

@Cpp.ExportAs("CppClass")
class MyClass {
   @Cpp.ExportAs("CppMember")
   var carbon_member: i32;
}
```

File: proposals/p2029.md [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+allowing undefined behaviour to occur.
```
Comment (zygoloid) [inline proposals/p2029.md:92]: Unless Carbon reserves a superset of all C++ reserved identifiers and keywords, we still have to deal with this issue. Moreover, the C++ keyword set will increase over time, so we will have to do something when that happens. For example, if a Carbon program uses the identifier `assume` and C++23 makes that a keyword, then we will have an interop conflict.

I am not convinced that we should be taking the approach of reserving all identifiers in Carbon that might be problematic for C++ interoperability. I think it's reasonable to instead tell Carbon developers to not use identifiers that can't be named from C++ in code that wants to interoperate with C++. There would be benefit in not reserving those identifiers, because the C++ restrictions do cause problems in practice: in particular, the restriction for identifiers beginning with underscore-capital and the restriction for identifiers containing a `__` not at the beginning affect code that people want to write reasonably often.

I think we should seriously consider the option of not having any reserved identifiers in Carbon, and see where that causes problems and how we might address those problems. For example:
- Can implementations provide non-standard keywords? How? (Maybe we should have a `k#foo` syntax for opting into keyword behavior, or maybe we should reserve a leading `__`. Explorer currently has some `__keyword`s for some experimental features that haven't been through the design proposal process.)
- How is built-in functionality exposed? (See #1804 for my thoughts on that.)
- Do we provide a mechanism to change the name of a Carbon entity when exposed to C++, and if so, how? (My preference here is to say no, and that if you want this behavior you write an alias on the Carbon side that will then be available in both languages -- or enable `-fms-extensions` in Clang and use `__identifier(whatever)` as the C++ analogue of Carbon's `r#whatever`.)
